### PR TITLE
Skip shipment creation for unreturned exchanges

### PR DESCRIPTION
**Pull request by richardnuno**
This change fixes a bug whereby the UnreturnedItemCharger would fail to create a new order when any of the variants in the new order was out of stock. Since the shipment has already shipped, no verification for stock quantity is required.

### CODE REVIEW

**Comment by jordan-brough** [comment]
:+1:

File: core/app/models/spree/order.rb [hunk 1]
```diff
@@ -516,6 +516,9 @@ def ensure_shipping_address
     end
 
     def create_proposed_shipments
+      # An unreturned exchange may have all of its shipments already shipped
+      return self.shipments if shipments.present? && shipments.all? { |s| s.shipped? }
```
Comment (jhawthorn) [inline core/app/models/spree/order.rb]: Is it possible to check that this is a "unreturned exchange" as well as doing this check?

I don't think that this "silent fail and skip" would make sense for a normal order.

File: core/spec/models/spree/order_spec.rb [hunk 2]
```diff
@@ -919,15 +919,23 @@ def compute(computable)
...
         expect {
           subject.create_proposed_shipments
         }.to raise_error(Spree::Order::CannotRebuildShipments)
       }.not_to change { subject.reload.shipments }
 
-      expect { shipment.reload }.not_to raise_error
+      expect { shipped_shipment.reload }.not_to raise_error
+    end
+
+    it "returns the shipments when all shipments are shipped" do
```
Comment (jhawthorn) [inline core/spec/models/spree/order_spec.rb]: This test tests the behaviour change, but doesn't expose why this change was necessary. Can we get a spec showing the problem with an unreturned exchange?

**Comment by richardnuno** [comment]
@jhawthorn updated with your feedback. Thanks for suggesting the extra spec, it actually uncovered another issue that would have happened further along while finalizing the order. When finalizing the new order, the shipment would be finalized again causing the count on hand for the variants in the shipment to be decremented (causing the out of stock error again if there is no stock). I've added a fix for that issue as well, mind taking another look?

**Comment by athal7** [comment]
:+1:

**Comment by richardnuno** [comment]
@athal7 updated with the symbol-to-proc syntax

**Comment by athal7** [comment]
looks great. do you mind squashing then we can merge when we get a review from FRT

**Comment by richardnuno** [comment]
@athal7 sure, squashed all the commits

**Comment by richardnuno** [comment]
@jhawthorn any chance you can take another look at this?

File: core/app/models/spree/order.rb [hunk 3]
```diff
@@ -516,6 +516,8 @@ def ensure_shipping_address
     end
 
     def create_proposed_shipments
+      return self.shipments if unreturned_exchange? && shipments.all?(&:shipped?)
```
Comment (jhawthorn) [inline core/app/models/spree/order.rb]: I think this should be a raise, possibly even a `raise ... if create_proposed_shipments`. This behaviour doesn't fit well with the rest of this methods behaviour. I think if we were to write documentation for this method it, this behaviour would seem really out of place.

If I change it to this to a `raise if unreturned_exchange? && shipments.all?(&:shipped?)` only the spec testing this exactly this method fails. Is it possible to get a spec which covers _why_ this behaviour is required rather than just that it exists

Comment (richardnuno) [inline core/app/models/spree/order.rb]: The spec that covers why it's required is in `unreturned_item_charger_spec.rb`. It's a though one to bring over to `order_spec.rb` since it's directly related to this method being called when attempting to charge for unreturned exchanges.

I think raising might be too severe since it's not an exceptional case, perhaps wrapping the contents of `create_proposed_shipments` within an `unless unreturned_exchange? && shipments.all?(&:shipped?)` would make it a little clearer?

Comment (jhawthorn) [inline core/app/models/spree/order.rb]: > The spec that covers why it's required is in unreturned_item_charger_spec.rb. It's a though one to bring over to order_spec.rb since it's directly related to this method being called when attempting to charge for unreturned exchanges.

I don't need it in order_spec, but nothing in that file shows that this is required. See jhawthorn/solidus@acc533053ac172550fea0bd2647ef4d91ca1234e I replace the return with a `raise "FOOBAR"` which never occurs (outside of the skipped unit test)

> I think raising might be too severe since it's not an exceptional case, perhaps wrapping the contents of create_proposed_shipments within an unless unreturned_exchange? && shipments.all?(&:shipped?) would make it a little clearer?

It is the behaviour, not the formatting that I think is unclear here. I think raising an exception is the right thing to do here: we've asked to create new shipments on an order that should not have a new shipment created. Raising an error informs the caller that we were not able to perform the requested action. Failing silently makes it impossible to determine what the result of the action was.

**Comment by richardnuno** [comment]
I took another look at the `unreturned_item_charger_spec` and you're right, it wasn't hitting the `create_proposed_shipments` at all. `set_shipment_for_new_order` was doing a `update_attributes!(state: "confirm")`, that I've since removed, causing the order to not transition to confirm via the state machine.

The code is calling `.next` on the order until the confirm step is reached or `.next` returns false. When a user has store credit, the order is able to reach the confirm step. When the customer doesn't, it would stay in the payment step `set_order_payment` would associate the payments and `set_shipment_for_new_order` would move the order to the confirm state by doing a `update_attributes`. Moving `set_shipment_for_new_order` up and removing the `update_attributes!(state: "confirm")` means that the order may not be in the confirm state before calling `.complete!` To fix this, I moved `set_order_payment` up before the attempt to transition the order to confirm. This means the order has all of the necessary associated data so that it's able to reach the confirm state.

Regarding the raise in the `create_proposed_shipments` (instead of returning the shipments), I agree it's something that shouldn't happen under normal circumstances but I also think the presence of a check like `unreturned_exchange? && shipments.all?(&:shipped?)` indicates that it's a special case we're trying to work around. Raising an error in `create_propsed_shipments` would leave the current bug unfixed. Definitely open to alternative solutions to fixing this bug but as we discussed over Slack a while ago, all of the other solutions we've come up with so far would also not be ideal (adding `unless: :unreturned_exchange?` to several state machine hooks) or very time consuming.

**Comment by jhawthorn** [comment]
@richardnuno I'm sorry to be a stickler, but we need a test for e3521f5bd380f0ec67b05f6c8a9c0950eb1d1fb4.

For all I know, this could have worked fine as it was before this commit, skipping the state machine using update_attributes! could have been a fine solution. If there's a reason it's not we need to show what that is.

I'm fine with the less than optimal change to `create_proposed_shipments` for the sake of fixing this bug, but we need to clearly show in tests why it has to be as it is. I would consider the unreturned exchanges one of the uglier parts of the codebase, and something I want to visit as soon as I have the time. In order for anyone to ever fix these things, we need the current behaviour and requirements to be well tested.

**Comment by richardnuno** [comment]
@jhawthorn totally reasonable request, just not sure how to go about testing that. I think the original intention was to have the order go through the state machine until completion. Since it didn't, the `update_attributes!(state: 'confirm')` remedied that but nothing points to intentionally wanting to skip steps in the state machine. Since I believe we always want the order to go through the state machine, I've added error raising whenever the order is not in the state I expect and tests for that in the latest commit.

Definitely want to add as much test coverage as possible but struggling to come up with good test cases. Could you be more specific about the tests you'd like to see in both the order and unreturned item charger specs?

**Comment by richardnuno** [comment]
Closing in favor of https://github.com/solidusio/solidus/pull/367
